ndb/src/google/cloud/ndb/model.py

```diff
@@ -2947,28 +2962,286 @@ def __init__(self, *args, **kwargs):
 
 
 class MetaModel(type):
-    __slots__ = ()
+    """Metaclass for Model.
 
-    def __new__(self, *args, **kwargs):
-        raise NotImplementedError
+    This exists to fix up the properties -- they need to know their name. For
+    example, defining a model:
 
+    .. code-block:: python
 
-class Model:
-    __slots__ = ("_entity_key",)
+        class Book(ndb.Model):
+            pages = ndb.IntegerProperty()
 
-    def __init__(self, *args, **kwargs):
-        raise NotImplementedError
+    the ``Book.pages`` property doesn't have the name ``pages`` assigned.
+    This is accomplished by calling the ``_fix_properties()`` method on the
+    class itself.
+    """
+
+    def __init__(cls, name, bases, classdict):
+        super(MetaModel, cls).__init__(name, bases, classdict)
+        cls._fix_up_properties()
+
+    def __repr__(cls):
+        props = []
+        for _, prop in sorted(cls._properties.items()):
+            props.append("{}={!r}".format(prop._code_name, prop))
+        return "{}<{}>".format(cls.__name__, ", ".join(props))
+
+
+class Model(metaclass=MetaModel):
+    """A class describing Cloud Datastore entities.
+
+    Model instances are usually called entities. All model classes
+    inheriting from :class:`Model` automatically have :class:`MetaModel` as
+    their metaclass, so that the properties are fixed up properly after the
+    class is defined.
+
+    Because of this, you cannot use the same :class:`Property` object to
+    describe multiple properties -- you must create separate :class:`Property`
+    objects for each property. For example, this does not work:
+
+    .. code-block:: python
+
+        reuse_prop = ndb.StringProperty()
+
+        class Wrong(ndb.Model):
+            first = reuse_prop
+            second = reuse_prop
+
+    instead each class attribute needs to be distinct:
+
+    .. code-block:: python
+
+        class NotWrong(ndb.Model):
+            first = ndb.StringProperty()
+            second = ndb.StringProperty()
+
+    The "kind" for a given :class:`Model` subclass is normally equal to the
+    class name (exclusive of the module name or any other parent scope). To
+    override the kind, define :meth:`_get_kind`, as follows:
+
+    .. code-block:: python
+
+        class MyModel(ndb.Model):
+            @classmethod
+            def _get_kind(cls):
+                return "AnotherKind"
+
+    A newly constructed entity will not be persisted to Cloud Datastore without
+    an explicit call to :meth:`put`.
+
+    .. note::
+
+        You cannot define a property named ``key``; the ``.key`` attribute
```

Comment:
AFAICT this is not explicitly enforced by the code

Reply:
I guess it's enforced by it just not working? Explicit enforcement with an informative exception raised might be better than whatever strange, broken behavior occurs naturally, although I don't know what the natural consequence is.

Reply:
I just ran this in a live app sandbox:

```python
from google.appengine.ext import ndb
class MyModel(ndb.Model):
    key = ndb.StringProperty()

m = MyModel(key="hi")
print(m)
m.put()
print(m)
```

It results in

```
MyModel(key='hi')
MyModel(key=Key('MyModel', 5690894162329600), key=u'hi')
```

Reply:
That's nonsense. Awesome.

Reply:
Yeah so I think the fix is to just remove the note?

Reply:
Is the above behavior not considered pathological? At best, it's at least confusing.

Reply:
I did a big revamp of this docstring with some doctests. I need to add tests (and a "fix") for `Model.__repr__` but I think it's much better. It may be worth rendering locally for you to get a picture?